# DATA_MINING: EXERCISE 1

In [1]:
import pandas as pd
import numpy as np
from itertools import combinations as C

#### PROBLEM 1.1

* Write a program which will find all such numbers which are symetric (such as 11, 121, 1221,...) between 10 and 2000 (both included). The numbers obtained should be printed in a comma separated sequence on a single line.

In [2]:
# We create a function to print the symetric number between 2 integer number a and b @-@
def symetric_num(a,b):                        # Note: a and b must be an integer !
    if (type(a)!= int) or (type(b)!= int):    
        print("PLease enter integer number @-@")
        
    set_of_num=np.linspace(a,b,b-a+1)
    
    for i in range(0,b-a+1):                  # We check whether the reverse of number is similar to that number *-*
        if str(int(set_of_num[i]))== str(int(set_of_num[i]))[::-1]:
            print (i, end=", ")
           

In [3]:
print("All symetric number between 10 and 2000 is :")
symetric_num(10,2000)

All symetric number between 10 and 2000 is :
1, 12, 23, 34, 45, 56, 67, 78, 89, 91, 101, 111, 121, 131, 141, 151, 161, 171, 181, 192, 202, 212, 222, 232, 242, 252, 262, 272, 282, 293, 303, 313, 323, 333, 343, 353, 363, 373, 383, 394, 404, 414, 424, 434, 444, 454, 464, 474, 484, 495, 505, 515, 525, 535, 545, 555, 565, 575, 585, 596, 606, 616, 626, 636, 646, 656, 666, 676, 686, 697, 707, 717, 727, 737, 747, 757, 767, 777, 787, 798, 808, 818, 828, 838, 848, 858, 868, 878, 888, 899, 909, 919, 929, 939, 949, 959, 969, 979, 989, 991, 1101, 1211, 1321, 1431, 1541, 1651, 1761, 1871, 1981, 

#### PROBLEM 1.2

* Write a program which will find all such numbers which include at least 3 duplicated digits (such
as 111, 1222,...) between 100 and 2000 (both included). The numbers obtained should be printed
in a comma-separated sequence on a single line.

In [4]:
# We create a function to identify all distinct digit in a number, count number of repetition and max of count @-@
# Actually, the problem is not this complicated, just for fun though *-*

def duplicated_digit(num):
    num_list=list(str(num))
    unique_value= np.unique(np.array(num_list))
    count_num =[]
    for i in unique_value:
        count_num.append(num_list.count(i))
    max_num=max(count_num)
    return_dict= {"distinct value":unique_value,"count":count_num, "max": max_num}
    return return_dict              
                

In [5]:
# For example, the number 1332447763399014:
duplicated_digit(1332447763399014)

{'distinct value': array(['0', '1', '2', '3', '4', '6', '7', '9'], dtype='<U1'),
 'count': [1, 2, 1, 4, 3, 1, 2, 2],
 'max': 4}

In [6]:
# So the number which repeaat at least 3 dupplicated digits when max_count (the most repeated digits) >=3:
for i in np.linspace(100,2000,1901):
    if duplicated_digit(int(i))["max"]>=3:
        print(int(i),end=", ")

111, 222, 333, 444, 555, 666, 777, 888, 999, 1000, 1011, 1101, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1121, 1131, 1141, 1151, 1161, 1171, 1181, 1191, 1211, 1222, 1311, 1333, 1411, 1444, 1511, 1555, 1611, 1666, 1711, 1777, 1811, 1888, 1911, 1999, 2000, 

#### PROBLEM 1.3

* Write a program which will receive an integer larger than 0 and respond number of digits of that
integer.

In [11]:
# Create a function to count number of digit of positive integer
def digit(num):
    if type(num)!=int or num<=0:
        print("PLease enter positive integer number")
    else:
        print("number of digit:",len(list(str(num))))

In [12]:
digit(-4)

PLease enter positive integer number


In [13]:
digit(34.5)

PLease enter positive integer number


In [14]:
digit(14443226)

number of digit: 8


#### PROBLEM 1.4

* Write a program which will receive an integer larger than 0 and present that integer in binary
format.

In [22]:
# Create a function to turn number to binary digits
def binary(num):
    if type(num)!=int or num<=0:
        print("PLease enter positive integer number")
    else:
        print("binary number:",format(num,"b")) 

In [23]:
binary(12.5)

PLease enter positive integer number


In [24]:
binary(-5)

PLease enter positive integer number


In [25]:
binary(1332554870023)

binary number: 10011011001000010011101011010110100000111


#### PROBLEM 1.5

* Write a program which will receive an integer larger than 100 and find all divisors of that integer.
The numbers obtained should be printed in a comma-separated sequence on a single line.

In [62]:
# Create a function to find all the divisor of positive integer >100:
def divisor(num):
    half_num=int(num/2)
    divisor=[1,num]
    for i in range(2,half_num+1):
            if (num%i)==0:
                divisor.append(i)
    divisor.sort()
    return divisor          

In [66]:
number=float(input("Enter positive integer larger than 100:"))
while number!= int(number) or number<=100:
    number=float(input("Please input again the postive integer >100: "))
number=int(number)

print("All divisors of number is: \n")
print(*divisor(number),sep=", ")

Enter positive integer larger than 100:2.6
Please input again the postive integer >100: 99
Please input again the postive integer >100: 128
All divisors of number is: 

1, 2, 4, 8, 16, 32, 64, 128


#### PROBLEM 1.6

* Write a program which will find all primes between 1000 and 3000 (both included). The numbers
obtained should be printed in a comma-separated sequence on a single line.

In [53]:
# Create a function to find number of divisor of an integer
def len_divisor(num):
    half_num=int(num/2)
    divisor=[1,num]
    for i in range(2,half_num+1):
            if (num%i)==0:
                divisor.append(i)
    return len(divisor)

In [57]:
# All prime numbers only have 2 divisors:
prime_num=[]
for i in range(1000,3001):
    if len_divisor(i)==2:
        prime_num.append(i)
print("all prime numbers are:\n", *prime_num, sep=", ")

all prime numbers are:
, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291, 1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451, 1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531, 1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601, 1607, 1609, 1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699, 1709, 1721, 1723, 1733, 1741, 1747, 1753, 1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889, 1901, 1907, 1913, 1931, 1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997, 1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083, 2087, 2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161, 2179, 2203, 2207, 2213, 222

#### PROBLEM 1.7

* Write a program which will receive an integer larger than 0 and respond all perfect numbers smaller
than that integer. A perfect number is sum of its divisors except itself, such as 6 = 1+2+3. The
numbers obtained should be printed in a comma-separated sequence on a single line.

In [73]:
# Create a function to find the perfect number
def perfect(num):
    divisor_num=divisor(num)
    if sum(divisor_num[0:len(divisor_num)-1])==num:
        return 1           # num is perfect number
    else:
        return 0           # num is not perfect number

In [76]:
number=float(input("Enter positive integer:"))
while number!= int(number):
    number=float(input("Please input again the postive integer: "))
number=int(number)

Enter positive integer:12.67
Please input again the postive integer: 8129


In [77]:
perfect_num=[]
for i in range(1,number):
    if perfect(i)==1:
        perfect_num.append(i)
        
print("All perfect numbers is: \n")
print(*perfect_num,sep=", ")

All perfect numbers is: 

1, 6, 28, 496, 8128


#### PROBLEM 1.8

* With a given integral number n, write a program to generate a dictionary that contains (i, i*i*i)
such that is an integral number between 1 and n (both included). and then the program should
print the dictionary. Suppose the following input is supplied to the program:
6
Then, the output should be:
{1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216}

In [81]:
# Create the dictionary:
def dictionary(n):
    dict={}
    key=range(1,n+1)
    for i in key:
        dict[i]=int(i)**3
    return dict

In [87]:
dictionary(20)

{1: 1,
 2: 8,
 3: 27,
 4: 64,
 5: 125,
 6: 216,
 7: 343,
 8: 512,
 9: 729,
 10: 1000,
 11: 1331,
 12: 1728,
 13: 2197,
 14: 2744,
 15: 3375,
 16: 4096,
 17: 4913,
 18: 5832,
 19: 6859,
 20: 8000}

#### PROBLEM 1.9

* Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input
and then check whether they are divisible by 3 or not. The numbers that are divisible by 3 are to
be printed in a comma separated sequence. Example:
0100,0011,1010,1101
Then the output should be:
1101


In [88]:
# Create a funcion to turn binary digit to integer number:
def binary_int(binary):
    integer=[]
    for i in range(len(binary)):
        integer.append(int(binary[i],2))
    return integer

In [99]:
binary=['0100','0011','1010','1101','1111','1011']
integer=binary_int(binary)
integer

[4, 3, 10, 13, 15, 11]

In [100]:
divisible=[]
for i in range(len(integer)):
    if not integer[i]%3:
        divisible.append(binary[i])

print("All binary numbers divisible by 3 is: \n")
print(*divisible,sep=", ")    

All binary numbers divisible by 3 is: 

0011, 1111


#### PROBLEM 1.10

* Write a program which will receive an integer larger than 0 and respond if that integer is sum of
three distinct squared numbers, such as 38 = 22 + 32 + 52.

In [2]:
# Create a function to return 1 if the input number is sum of 3 distinct squared numbers, and return 0 otherwisse
def sum_of_square(number):
    square=[]
    for i in range(1,(int(number**(1/2)))+1):
        square.append(i**2)
    combi_list=list(C(square,3))
    sum_n=[x+y+z for (x,y,z) in combi_list]
    n=0
    for i in range(len(sum_n)):
        if sum_n[i]==number:
            n=n+1
    return n

In [3]:
number=float(input("Enter number: "))
while number!= int(number) or number<= 0:
    number=float(input("Please input again the postive integer >0: "))
number=int(number)

if sum_of_square(number)==1:
    print("The input number is sum of 3 distinct square numbers")
else:
    print("The input number is sum of 3 distinct square numbers")

Enter number: -2
Please input again the postive integer >0: 32.5
Please input again the postive integer >0: 38
The input number is sum of 3 distinct square numbers
